### ISSUE TITLE
perf: sequence models, speed up feedback loop (filter on `__get_item__`?)

### ISSUE THREAD

**Issue by MartinBernstorff** [Aarhus-Psychiatry-Research/psycop-common#545]
Avoids long waits filtering the dataset before starting training. This is a really-nice-to-have, but of course not mandatory.

### Filter in `__getitem__`
The `__getitem__` method of the dataset returns values at a given index. It seems iffy to filter at that step, because how do we handle the item at the index being filtered? Return a random item? Return `None`? And how does the `pl.Trainer` handle that?

We could return a random item, but that seems incredibly hacky.

I considered a pytorch `IterableDataset`, but then a `DataLoader` cannot shuffle it natively, we would have to implement that ourselves. 

* We could, of course, just shuffle before adding it to the dataloader, which seems like a reasonable choice?

### Move the mapping outside of the DataLoader
Mapping and filtering large amounts of data is exactly what Polars is really good at. Perhaps we want to move it somewhere before the `PatientSlice` is initialised instead? Current flow is roughly:
```
CohortDefiner --PolarsFrame--> 
CohortToPredictionTimes --PredictionTime[PatientSlice, Label]--> 
model.filter_and_reformat --PatientSlice--> 
embedder
```

* Perhaps we can move the heavy lifting to a subclass of `CohortToPredictionTimes`, and then have the `embedder` validate the input (e.g. that all tokens are within its vocabulary)?

**Comment by MartinBernstorff** [comment]
@KennethEnevoldsen What do you think? Feel free to just read through it, and then we can talk about it tomorrow.

**Comment by KennethEnevoldsen** [comment]
I believe before any of this we should probably look at whether we can simply speed up the filtering that is there, but to address some of the major points:

> We could, of course, just shuffle before adding it to the dataloader, which seems like a reasonable choice?

Not possible as this would lead to the same shuffling every epoch. If we train for multiple epochs this becomes a problem. We could fix that, but would much rather not have to do that.

> Move the mapping outside of the DataLoader

I don't believe it is a problem that the mapping is in the dataloader. It is the filtering before and I def. feel like that should work on patient slices.

**Comment by KennethEnevoldsen** [comment]
let us talk about this one
